### PR TITLE
Fix Extrapolation

### PR DESCRIPTION
**Pull request by SimonDanisch**
Combine the approaches from #15 #12 and adds comparison tests with PyMNE.
Still needs some clean up of the API and documentation + comments.

### CODE REVIEW

**Comment by SimonDanisch** [comment]
Ugh, I knew that a python test dependency would be trouble^^
```
The Python package matplotlib could not be imported by pyimport. Usually this means
that you did not install matplotlib in the Python version being used by PyCall.

PyCall is currently configured to use the Python version at:

/usr/bin/python3

and you should use whatever mechanism you usually use (apt-get, pip, conda,
etcetera) to install the Python package containing the matplotlib module.
```

**Comment by SimonDanisch** [comment]
Now, this adds a new geometry extrapolation, which adds points from a geometry (right now Rect + Circle), and then uses `ScatteredInterpolation` with Shepard distance (basically just a mean weighted by the distances to the existing points) to give those extra points values.
The code is here https://github.com/MakieOrg/TopoPlots.jl/blob/e51420abda16e15870e92b24e4133dd838033c19/src/extrapolation.jl#L41 (linked the likely most problematic line).

I also added some tests for creating those points:
Red: Bounding geometry
Blue: Extended geometry to minimize interference (took the 3x blow up from mne):
image
image

I also added new tests to directly compare the output to MNE.
Format is [TopoPlots.jl  vs PyMNE] and if the scatter marker are very visible, it means a peak shift... I think we now do pretty well!
Circular extrapolation: 
image
Rectangular extrapolation (like mne):
image

Problematic peaks:
image

Standard data:
image

**Comment by behinger** [comment]
wow! looks really good. Very impressed about the integration of matplotlib plots...

 I have a quick question: for me ScatteredInterpolation.jl was super slow (#5). Did you observe the same, or is it negligible because you only interpolate few points?

**Comment by SimonDanisch** [comment]
Ah come on...why is there always a different error with the python deps every time there's a new CI run?
image

**Comment by SimonDanisch** [comment]
@behinger good question... 
I haven't benchmarked it, but I was pretty sure that it should easily be fast enough for the few points, since it just calculates a couple of distances. 
I imagine that it would be quite different for a 512x512 matrix of points to interpolate, which should be fairly slow!

**Comment by SimonDanisch** [comment]
Added a graphic to the docs to explain the extrapolation:


test

File: Project.toml [hunk 1]
```diff
@@ -12,19 +12,19 @@ LinearAlgebra = "37e2e46d-f89d-539d-b4ee-838fcccc9c8e"
 Makie = "ee78f7c6-11fb-53f2-987a-cfe4a2b5a57a"
 Parameters = "d96e819e-fc66-5662-9728-84c9c7592b0a"
 PyCall = "438e738f-606a-5dbb-bf0a-cddfbfd45ab0"
+ScatteredInterpolation = "3f865c0f-6dca-5f4d-999b-29fe1e7e3c92"
```
Comment (palday) [inline Project.toml:15]: missing compat entry

File: test/runtests.jl [hunk 2]
```diff
@@ -1,12 +1,58 @@
...
-using LinearAlgebra, Statistics, TopoPlots, CairoMakie
+using LinearAlgebra, Statistics, TopoPlots, CairoMakie, FileIO
+using PyCall
+try
+    PyCall.pyimport("mne")
+catch e
+    # Hmpf, how else can we make sure a test dependency exists?
+    # I tried adding Conda for PyPlot, which then installs matplotlib automatically.
+    # It looks like this messed with mne, so one of those will need to have an ugly install?!
+    run(PyCall.python_cmd(`-m pip install mne`))
```
Comment (palday) [inline test/runtests.jl:10]: The build step of PyMNE 0.1.x should do this for you

Comment (SimonDanisch) [reply test/runtests.jl:10]: Well it doesn't somehow, since I added Conda to the test deps...

Comment (palday) [reply test/runtests.jl:10]: Conda is changing which Python is used after PyMNE has already been built.

File: src/core-recipe.jl [hunk 3]
```diff
@@ -65,48 +68,75 @@ macro plot_or_defaults(var, defaults)
...
-        return points_padded
+
+    geometry = lift(p.bounding_geometry, positions, p.enlarge; ignore_equal_values=true) do bounding_geometry, positions, enlarge
+        if bounding_geometry isa Type{<:GeometryPrimitive}
+            return enclosing_geometry(bounding_geometry, positions, enlarge)
+        elseif bounding_geometry isa GeometryPrimitive
+            return bounding_geometry
+        else
+            error("Wrong type for `bounding_geometry`: $(bounding_geometry)")
+        end
```
Comment (palday) [inline src/core-recipe.jl:84]: This feels like a place for using multiple dispatch.....

File: src/core-recipe.jl [hunk 4]
```diff
@@ -65,48 +68,75 @@ macro plot_or_defaults(var, defaults)
...
         yg[] = LinRange(ymin, ymax, resolution[2])
         return
     end
+
     notify(p.resolution) # trigger above (we really need `update=true` for onany)
 
-    padded_data = lift(pad_data, p.data, npositions, p.pad_value)
+    p.geometry = geometry # store geometry in plot object, so others can access it
+
+
```
Comment (palday) [inline src/core-recipe.jl:101]:
```suggestion
```

File: test/runtests.jl [hunk 5]
```diff
@@ -1,12 +1,59 @@
...
+PyPlot.pygui(false)
 
 include("percy.jl")
 
 data, positions = TopoPlots.example_data()
 
+function mne_topoplot(fig, data, positions)
+    circle = TopoPlots.enclosing_geometry(Circle, positions)
+    # Seems like the only way to plot positions with matching head is to norm the positions..
+    # Anything else leads to anarchy!
```
Comment (palday) [inline test/runtests.jl:23]: how does this relate to #18?

File: test/test-scripts.jl [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+    positions = rand(Point2f, 10)
+    data = rand(10)
+    rect = Rect(positions)
+    pos_extra, rect_extended, data_extra = TopoPlots.extrapolate_data(rect, positions, data)
+    f, ax, p = Makie.scatter(pos_extra, color=data_extra, markersize=10, axis=(aspect=DataAspect(),))
+    scatter!(ax, positions)
+    lines!(ax, rect)
+    lines!(ax, rect_extended, color=:red)
+    f
+end
```
Comment (palday) [inline test/test-scripts.jl:90]: did you mean to check in `test.png`?

File: test/runtests.jl (hunk 5 above)
Comment (SimonDanisch) [reply test/runtests.jl:23]: Good question... I don't think it does, but I don't know MNE well enough, to know what's going wrong on there side.
I'll take a look at #18 after this PR!

File: test/test-scripts.jl (hunk 6 above)
Comment (SimonDanisch) [reply test/test-scripts.jl:90]: Nope :D

**Review by palday — APPROVED** [review]

**Pull request merged**